planner/core/logical_plans.go

```diff
@@ -896,6 +903,27 @@ type FrameBound struct {
 	CmpFuncs []expression.CompareFunc
 }
 
+// Equal checks whether two FrameBounds are equal.
+func (bound *FrameBound) Equal(ctx sessionctx.Context, other *FrameBound) bool {
+	if len(bound.CalcFuncs) != len(other.CalcFuncs) || len(bound.CmpFuncs) != len(other.CmpFuncs) {
+		return false
+	}
+	for i := range bound.CalcFuncs {
+		if !bound.CalcFuncs[i].Equal(ctx, other.CalcFuncs[i]) {
+			return false
+		}
+	}
+	for i := range bound.CmpFuncs {
+		v1 := reflect.ValueOf(&bound.CmpFuncs[i]).Elem()
+		v2 := reflect.ValueOf(&other.CmpFuncs[i]).Elem()
+		if v1.Pointer() != v2.Pointer() {
+			return false
+		}
+	}
+	return bound.Type == other.Type && bound.UnBounded == other.UnBounded &&
```

Comment:
Could we check these field at first? Because the comparison of `CmpFuncs` is really expensive.